lib/ws-channel.js

```diff
@@ -0,0 +1,252 @@
+
+module.exports = WebSocketChannel;
+
+var assert = require('assert');
+var crypto = require('crypto');
+var debug = require('./debug')('ws-channel');
+var events = require('events');
+var extend = require('util')._extend;
+var util = require('util');
+var WebSocket = require('ws');
+
+var REQ = 'strong-control-channel:request';
+var RSP = 'strong-control-channel:response';
+var NOT = 'strong-control-channel:notification';
+
+
+function WebSocketChannel(onRequest, onNotification) {
+  assert(onRequest);
+  assert(onNotification);
+
+  this._onRequest = onRequest;
+  this._onNotification = onNotification;
+
+  this._websocket = null;
+  this._socket = null;
+
+  // Initialize to empty.
+  this._error = null;
+  this._sendQueue = [];
+  this._uri = null;
+  this._callbacks = {};
+  this._callbacksPending = 0;
+  this._seqno = 0;
+  this._token = null;
+  this._unref = false;
+}
+
+util.inherits(WebSocketChannel, events.EventEmitter);
+
+
+WebSocketChannel.prototype.connect = function connect(uri, token) {
+  assert(!this._websocket);
+  assert(uri);
+  assert(token);
+
+  this._uri = uri;
+  this._token = token;
+
+  var headers = { 'x-mesh-token': token };
+  var websocket = new WebSocket(uri, { headers: headers });
+
+  this._attach(websocket);
+
+  return this;
+};
+
+
+WebSocketChannel.prototype.createToken = function createToken() {
+  assert(!this._token);
+  return this._token = crypto.randomBytes(24).toString('hex');
+};
+
+
+WebSocketChannel.prototype.accept = function accept(websocket) {
+  assert(!this._websocket);
+  assert(this._token);
+
+  this._uri = null; // TODO(bert) set this
+  this._attach(websocket);
+
+  return this;
+};
+
+
+WebSocketChannel.prototype._attach = function _attach(websocket) {
+  this._websocket = websocket;
+
+  websocket.on('message', this._handleMessage.bind(this));
+  websocket.on('error', this._handleError.bind(this));
+
+  if (websocket.readyState === WebSocket.CONNECTING)
+    websocket.once('open', this._onOpen.bind(this));
+  else
+    this._onOpen();
+
+  this._flushSendQueue();
+};
+
+
+WebSocketChannel.prototype.close = function close(callback) {
+  if (this._websocket)
+    this._websocket.close();
+
+  if (this._sendQueue.length === 0) {
+    process.nextTick(callback);
+
+  } else {
+    process.nextTick(function() {
+      callback(new Error('Messages were discarded'));
```

Comment:
General node question.. wouldn't it be more useful to the developer if this Error was created before the `nextTick()` so that the stacktrace doesn't just show `nextTick`?